src/main/java/bio/terra/service/snapshot/SnapshotDao.java

```diff
@@ -432,9 +433,12 @@ private SnapshotProject retrieveSnapshotProject(String sql, MapSqlParameterSourc
 
         params.addValue("offset", offset).addValue("limit", limit);
         List<SnapshotSummary> summaries = jdbcTemplate.query(sql, params, new SnapshotSummaryMapper());
-
+        List<SnapshotSummary> summariesWithSources = summaries
```

Comment:
I worry that this might end up hammering the database since I think retrieveSnapshot is expensive.  Do you think it's possible to put the region extraction logic right in the SQL?

Reply:
Yeah that's a good point!

Reply:
I was able to get all of the storage information with this query. Is there a more optimal way you can think of?

```
SELECT snapshot_source.id, 
sr1.cloud_resource as cloud_resource1, sr1.region as region1, sr1.cloud_platform as cloud_platform1,
sr2.cloud_resource as cloud_resource2, sr2.region as region2, sr2.cloud_platform as cloud_platform2,
sr3.cloud_resource as cloud_resource3, sr3.region as region3, sr3.cloud_platform as cloud_platform3
FROM snapshot JOIN snapshot_source ON snapshot.id = snapshot_source.snapshot_id 
LEFT JOIN storage_resource as sr1 ON sr1.cloud_resource = 'BUCKET' AND sr1.dataset_id = snapshot_source.dataset_id 
LEFT JOIN storage_resource as sr2 ON sr2.cloud_resource = 'FIRESTORE' AND sr2.dataset_id = snapshot_source.dataset_id 
LEFT JOIN storage_resource as sr3 ON sr3.cloud_resource = 'BIGQUERY' AND sr3.dataset_id = snapshot_source.dataset_id;
```

Reply:
I don't know how comfortable we feel about this but we could totally do something with the json function in postgres to have this return the storage object in a single json string within the result set. It could look some thing like:
```
SELECT snapshot_source.id,
       (SELECT jsonb_agg(sr)
          FROM (SELECT region, cloud_resource as "cloudResource", cloud_platform as "cloudPlatform"
                  FROM storage_resource
                 WHERE dataset_id = snapshot_source.dataset_id) sr) AS storage
FROM snapshot JOIN snapshot_source  ON snapshot.id = snapshot_source.snapshot_id;
```

That returns stuff like in the storage column like:
```
[{"region":"US","cloudResource":"BIGQUERY","cloudPlatform":"GCP"}, 
 {"region":"US_CENTRAL1","cloudResource":"BUCKET","cloudPlatform":"GCP"}, 
 {"region":"US_CENTRAL1","cloudResource":"FIRESTORE","cloudPlatform":"GCP"}]
```

You might need to play with the mapping but this would be dynamic based on the key values in the table.  Lemme know what you think.  The JSON functions world in Postgres is wild!

Reply:
Woah that's wild!! I like this option since we won't need to update the sql query when we add azure storage.

Reply:
This is neat but isn't it awkward to work with JSON data in Java? Maybe `array_agg` would work better, I think jdbc can map its result to a collection type.

Reply:
I tried changing this to `array_agg` and it doesn't include the column names. It also seems awkward to rely on the order of the columns from the query to do the mapping 😞: 
`{"(US_CENTRAL1,BIGQUERY,GCP)","(US_CENTRAL1,FIRESTORE,GCP)","(US_CENTRAL1,BUCKET,GCP)"}`

Reply:
I wonder if the JDBC driver has anything to map the JSON to a Pojo.  IIRC, I think it returns it as a string so we can single use `objectMapper.readValue(rs.getString("storage"), Storage.class)`

Reply:
But yeah, I've relied pretty heavily on JSON in postgres in the past.  It can lead to some much more efficient querying, weirdly enough!  You can get a super nice mix of relational + noSQL

Reply:
IMO there's no problem depending on ordering as the `SELECT region, cloud_resource, cloud_platform` makes the ordering explicit. But either way should be fine. Json performance is poor but I doubt it would be an issue unless we have 1000s of elements in the response

I'm pretty sure there's a pure SQL way to do this as well, where the other records are denormalized across snapshot IDs. You could loop over all the results and use a `Map` to regain the snapshot ID collection relationship. SQL has many many ways to do the same thing.